Fixed #31358 -- Increased salt entropy of password hashers.

darakian opened the pull request:
The choice of 12 characters (~71bit) seems to stem from 2011 at latest
and may be based off the guidance in rfc 2898 circa 2000 which recommends at least 64 bits
https://www.ietf.org/rfc/rfc2898.txt
The git commit containing the original code is
dce820ff70f00e974afd3e6e310aa825bc55319f

Modern guidance suggests the use of at least 128 bits for salt values
OWASP Guidance: https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Password_Storage_Cheat_Sheet.md#salting
Python guidance for pbkdf2: https://docs.python.org/3/library/hashlib.html#hashlib.pbkdf2_hmac
NIST Guidance: https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-132.pdf See: Section 5.1

In the nist case this is actually a hard requirement which could be placing some users out of compliance.

This PR bumps the default length of get_random_string() to 22 which provides approximately 131 bits of entropy
log_2((26+26+10)^22) = 130.9923...

felixxm commented:
@darakian Thanks for this patch :+1: Please check comments in the ticket-31358. It would also be good to add a module level constant, e.g. `SALT_LENGTH = 22` in the `contrib/auth/hashers.py`.

darakian commented:
Hey @felixxm. Glad to contribute :)
I've rebased on the latest master to include the new version of `get_random_string` and this branch now just changes the the call to `get_random_string`.

darakian commented:
@felixxm I've updated this branch as per comments on https://github.com/django/django/pull/13107
The commit from that branch is now the first commit on this branch though I've added and squashed in the change requested on removing the cached property. I've also squashed down all of the older commits on from this branch and have redone them to work in terms of bits rather than in terms of character lengths.

darakian commented:
Bit more git noise and tests are passing. Turns out I goofed on `must_update` with
```
return len(self.safe_summary(encoded)['salt']) >= self.salt_entropy / math.log(62, 2)
```
which of course is exactly wrong. I've updated the `>=` to `<`.  Math is hard. 🤦

felixxm commented on django/contrib/auth/hashers.py:
```diff
@@ -228,7 +231,11 @@ def safe_summary(self, encoded):
-        return False
+        try:
+            # A string of length: N / log_2(26+26+10), provides N bits of entropy.
+            return len(self.safe_summary(encoded)['salt']) < self.salt_entropy / math.log(62, 2)
```
I'm starting to think if this is an expected change :thinking: This can force updates each time when the password is checked for password created with a `salt` passed manually to the `make_password()`.

apollo13 replied:
Good question. It will ensure that salt_entropy is met… which brings me back to the comment mentioned in the ticket: Does it make sense to support a `salt` for encode at all? You need to know the hasher to be able to provide the proper `salt` (think the bcrypt format which encodes rounds in it).

Sadly this will be a rather tricky issue to work on. I also think that `must_update` shouldn't do this calculation of the `salt_len` but rather the hashers should return it from `decode` so they are free to store it in a different form…

darakian replied:
If the hashers should each be making their own determination then should I open another PR to move `must_update` to be a required function for each hasher?

apollo13 replied:
Mhm that is what I was trying to avoid, because for most hasher a salt length is just that and `must_update` should easily be able to handle that globally if it is returned from `decode`. 

What this PR certainly misses (and what will show you the existing problems) is a test for the behavior of the `bcrypt` hasher. I think now it's `must_update` will *always* return `True` and set a salt *every* time.

darakian replied:
I'm not sure I see the issue. `BCryptSHA256PasswordHasher` defines its own `must_update` as
```
    def must_update(self, encoded):
        decoded = self.decode(encoded)
        return decoded['work_factor'] != self.rounds
```
and `BCryptPasswordHasher` inherits from it. Maybe bcrypt should take the salt into account in that function, but I don't see how it would always return true. I've added some tests for this as well. 

That said, I've noticed that most of the hashers have a custom `must_update` which does not account for the salt size. I think they should and if you agree then I'll add logic to them to check.

darakian replied:
@apollo13 Where do you stand on the `must_update` change? Is that something that should be addressed later? If so what would be the proper channel in which to raise this topic for discussion?

darakian replied:
@apollo13 Can we circle back to this? I've added a salt strength check to the pbkdf2 hasher as well as two tests related to checking the behavior of `must_update`. Maybe pbkdf2 is the only hasher that it makes sense to add this check to.

apollo13 replied:
> Maybe pbkdf2 is the only hasher that it makes sense to add this check to.

Why? It certainly also makes sense for Argon and the standard salted sha/md5 stuff. The main problem I see currently is that https://github.com/django/django/blob/d618712a3b9df95509a053cf9fa444f47d4d0c79/django/contrib/auth/hashers.py#L236 is simply wrong for some hashers, so yeah maybe we cannot provide a generic `must_update` method (ie the crypt hasher has a salt length of 2 [literally]).

> That said, I've noticed that most of the hashers have a custom must_update which do not account for the salt size. I think they should account for salt size and if you agree then I'll add logic to them to check.

Yes my initial idea was that all hasher you call `must_update` that would allow them to determine if salt changes are needed. Sadly I do not think it is possible to write a `salt` check in a generic way.

To be honest I do not have good ideas about that. For this to be mergable we'd at least have to have it working for Argon & pbkdf2 (sha/md5 is easy then) without to many hacks / code smells.

apollo13 replied:
Also what happens if one passes in a salt manually? ie literally `hasher.encode(self, password, salt="123",…)` how can we ever enforce an entropy there?

darakian replied:
> Yes my initial idea was that all hasher you call must_update that would allow them to determine if salt changes are needed. Sadly I do not think it is possible to write a salt check in a generic way.

> To be honest I do not have good ideas about that. For this to be mergable we'd at least have to have it working for Argon & pbkdf2 (sha/md5 is easy then) without to many hacks / code smells.

I agree. I think this will need to be hasher specific though the base password hasher can provide a default case. I hadn't considered manual salt overrides and that's a good question. Generally I think we need to have memory of a manual override which we do not have today, so we either add memory to the stored hash somehow or we require the caller to remember and allow them to override the entropy value used in `must_update`. I don't think either solution is very clean.

As for the hashers; the salted md5/sha hashers don't implement their own `must_update` so they're already covered by the base password hasher implementation. I've added a salt check to argon2 as well with a minor simplification that it does not check for no salt given that the rest of the check also assumes a salt exists.

darakian replied:
> Yes my initial idea was that all hasher you call must_update that would allow them to determine if salt changes are needed. Sadly I do not think it is possible to write a salt check in a generic way.

> To be honest I do not have good ideas about that. For this to be mergable we'd at least have to have it working for Argon & pbkdf2 (sha/md5 is easy then) without to many hacks / code smells.

I agree. I think this will need to be hasher specific. I hadn't considered manual salt overrides and that's a good question. Generally I think we need to have memory of a manual override which we do not have today, so we either add memory to the stored hash somehow or we require the caller to remember and allow them to override the entropy value used in `must_update`. I don't think either solution is very clean.

As for the hashers; the salted md5/sha hashers don't implement their own `must_update` so they're already covered by the base password hasher implementation. I've added a salt check to md5/sha1 and argon2. Argon2 has a minor simplification that it does not check for the existence of the salt given that the rest of the check also assumes a salt exists. For md5 and sha1 this add `must_update` to each.

darakian replied:
@apollo13 any thoughts on the manual salting question? Also I've squashed a few of my commits down to clean up the salt strength checking portion of this PR.

darakian replied:
@apollo13 Alternately, how do you feel about breaking out the update portion of this PR? I can make a separate PR to handle the discussion of updating and reduce this PR to change the measurement of the salt from characters to bits and to set the default to 128 bits.

darakian commented on django/contrib/auth/hashers.py:
```diff
@@ -596,7 +603,7 @@ def decode(self, encoded):
             'algorithm': self.algorithm,
             'hash': encoded[6:],
-            'salt': None,
+            'salt': '',
```
Ok updated.

darakian commented on django/contrib/auth/hashers.py:
```diff
@@ -190,8 +192,9 @@ def _load_library(self):
-        # 12 returns a 71-bit value, log_2((26+26+10)^12) =~ 71 bits
-        return get_random_string(12)
+        # Salt hash with entropy of at least salt_entropy bits
+        char_count = math.ceil(self.salt_entropy / math.log(62, 2))
```
Certainly a detail I missed. I've updated the PR, but if you have any links describing this I'd love to read about it.

> I suggest unchecking the "Needs..." flags on the ticket if that's true to make this more visible for reviewers.

Updated. Thanks for the tip.

jacobtylerwalls replied:
Nothing fancy, it's just mentioned right in the Python module documentation: https://docs.python.org/3.3/library/math.html#module-math

apollo13 commented:
I have been thinking long and hard about this PR and it's implications. I really hate the code duplication and magic numbers it introduces. The latter is easily fixable by moving 62 etc into a constant and adding a comment to it. The code duplication could be eliminated by moving the common code into an function like `salt_length_changed`.

Another option would be to consider the salt length a "system" property and as such test for changes here:
https://github.com/django/django/blob/429d089d0a8fbd400e0c010708df4f0d16218970/django/contrib/auth/hashers.py#L48-L49

This still would need an extra attribute on the hashers ala `supports_salt_change`, so dunno.

I think all in all (for simplicity) I am leaning towards just cleaning up the existing code and shipping it in 3.2. I'll push a commit to that extend later today into this PR.

apollo13 commented on django/contrib/auth/hashers.py:
```diff
@@ -384,10 +390,12 @@ def must_update(self, encoded):
         current_params = argon2.extract_parameters('$' + rest)
         new_params = self.params()
+        # A string of length: N / log_2(26+26+10), provides N bits of entropy.
+        weak_salt = current_params.salt_len < self.salt_entropy / math.log(62, 2)
```
While trying to push this PR over the finish line I noted that `salt_len` is not really the `salt_len` we want (it's the size of a string in bytes). I have pushed a new PR #13815 which adds the actual salt to `decode`. I'll update this PR once the other one is merged.

darakian commented:
> I have been thinking long and hard about this PR and it's implications. I really hate the code duplication and magic numbers it introduces. The latter is easily fixable by moving 62 etc into a constant and adding a comment to it. The code duplication could be eliminated by moving the common code into an function like `salt_length_changed`.
> 

I've moved 62 out to a named variable. As for the code duplication; are you referring to the `must_decode` functions?

apollo13 commented:
Yes, but I had a somewhat simpler fix in mind:

```
diff --git a/django/contrib/auth/hashers.py b/django/contrib/auth/hashers.py
index 214618b42d..bb1081a376 100644
--- a/django/contrib/auth/hashers.py
+++ b/django/contrib/auth/hashers.py
@@ -11,7 +11,7 @@ from django.core.exceptions import ImproperlyConfigured
 from django.core.signals import setting_changed
 from django.dispatch import receiver
 from django.utils.crypto import (
-    constant_time_compare, get_random_string, pbkdf2,
+    RANDOM_STRING_CHARS, constant_time_compare, get_random_string, pbkdf2,
 )
 from django.utils.module_loading import import_string
 from django.utils.translation import gettext_noop as _
@@ -162,6 +162,11 @@ def mask_hash(hash, show=6, char="*"):
     return masked
 
 
+def must_update_salt(salt, wanted_salt_entropy):
+    # Each character in the salt provides log_2(len(alphabet)) bits of entropy
+    return len(salt) * math.log2(len(RANDOM_STRING_CHARS)) < wanted_salt_entropy
+
+
 class BasePasswordHasher:
     """
     Abstract base class for password hashers
@@ -193,8 +198,8 @@ class BasePasswordHasher:
     def salt(self):
         """Generate a cryptographically secure nonce salt in ASCII."""
         # Salt hash with entropy of at least salt_entropy bits
-        char_count = math.ceil(self.salt_entropy / math.log2(62))
-        return get_random_string(char_count)
+        char_count = math.ceil(self.salt_entropy / math.log2(len(RANDOM_STRING_CHARS)))
+        return get_random_string(char_count, RANDOM_STRING_CHARS)
 
     def verify(self, password, encoded):
         """Check if the given password is correct."""
@@ -293,10 +298,8 @@ class PBKDF2PasswordHasher(BasePasswordHasher):
 
     def must_update(self, encoded):
         decoded = self.decode(encoded)
-        salt = decoded['salt']
-        # A string of length: N / log_2(26+26+10), provides N bits of entropy.
-        weak_salt = len(salt) < self.salt_entropy / math.log(62, 2) if salt is not None else False
-        return (decoded['iterations'] != self.iterations) or weak_salt
+        update_salt = must_update_salt(decoded['salt'], self.salt_entropy)
+        return (decoded['iterations'] != self.iterations) or update_salt
 
     def harden_runtime(self, password, encoded):
         decoded = self.decode(encoded)
@@ -390,12 +393,12 @@ class Argon2PasswordHasher(BasePasswordHasher):
         argon2 = self._load_library()
         current_params = argon2.extract_parameters('$' + rest)
         new_params = self.params()
-        # A string of length: N / log_2(26+26+10), provides N bits of entropy.
-        weak_salt = current_params.salt_len < self.salt_entropy / math.log(62, 2)
         # Set salt_len to the salt_len of the current parameters because salt
         # is explicitly passed to argon2.
         new_params.salt_len = current_params.salt_len
-        return (current_params != new_params) or weak_salt
+        update_salt = must_update_salt(current_params.salt, self.salt_entropy)
+        raise
+        return (current_params != new_params) or update_salt
 
     def harden_runtime(self, password, encoded):
         # The runtime for Argon2 is too complicated to implement a sensible
@@ -540,10 +543,7 @@ class SHA1PasswordHasher(BasePasswordHasher):
 
     def must_update(self, encoded):
         decoded = self.decode(encoded)
-        salt = decoded['salt']
-        # A string of length: N / log_2(26+26+10), provides N bits of entropy.
-        weak_salt = len(salt) < self.salt_entropy / math.log(62, 2) if salt is not None else False
-        return weak_salt
+        return must_update_salt(decoded['salt'], self.salt_entropy)
 
     def harden_runtime(self, password, encoded):
         pass
@@ -585,10 +585,7 @@ class MD5PasswordHasher(BasePasswordHasher):
 
     def must_update(self, encoded):
         decoded = self.decode(encoded)
-        salt = decoded['salt']
-        # A string of length: N / log_2(26+26+10), provides N bits of entropy.
-        weak_salt = len(salt) < self.salt_entropy / math.log(62, 2) if salt is not None else False
-        return weak_salt
+        return must_update_salt(decoded['salt'], self.salt_entropy)
 
     def harden_runtime(self, password, encoded):
         pass
diff --git a/django/utils/crypto.py b/django/utils/crypto.py
index db19db91d3..fb076104ee 100644
--- a/django/utils/crypto.py
+++ b/django/utils/crypto.py
@@ -47,14 +47,12 @@ def salted_hmac(key_salt, value, secret=None, *, algorithm='sha1'):
 
 
 NOT_PROVIDED = object()  # RemovedInDjango40Warning.
+RANDOM_STRING_CHARS = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'
 
 
 # RemovedInDjango40Warning: when the deprecation ends, replace with:
 #   def get_random_string(length, allowed_chars='...'):
-def get_random_string(length=NOT_PROVIDED, allowed_chars=(
-    'abcdefghijklmnopqrstuvwxyz'
-    'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'
-)):
+def get_random_string(length=NOT_PROVIDED, allowed_chars=RANDOM_STRING_CHARS):
     """
     Return a securely generated random string.
 
```

It still depends on #13815 first. Please do not update the PR further for now to prevent duplicate work. I hope to get this commited over the next week.

darakian commented:
> It still depends on #13815 first. Please do not update the PR further for now to prevent duplicate work. I hope to get this commited over the next week.

Sounds good and I'm fine with that change. I'll leave it as is. Thanks for getting back to this as well 👍

apollo13 commented:
@darakian I have (force) pushed the updates to your branch after rebasing. What do you think?

@felixxm I think we reached a stage where you can do a final review. Feel free to be very pedantic on this one; after all it touches the core of our password hashing ;)

darakian commented:
> @darakian I have (force) pushed the updates to your branch after rebasing. What do you think?

Looks good to me. 👍

felixxm reviewed (commented):
Thanks both :rocket: I added release notes and pushed small edits.

carltongibson reviewed (approved):
Looks good, thank you all.

apollo13 commented on django/contrib/auth/hashers.py:
```diff
@@ -189,9 +196,14 @@ def _load_library(self):
+        of at least `salt_entropy` bits.
+        """
+        # Each character in the salt provides log_2(len(alphabet)) bits of
+        # entropy.
```
Really really minor nitpick; if flake8 etc do not complain maybe move `entropy.` up one line; looks really ugly like this especially if the code below is longer :)

felixxm replied:
I cannot move it to the same line per our [coding style](https://docs.djangoproject.com/en/3.1/internals/contributing/writing-code/coding-style/#python-style): _"Documentation, comments, and docstrings should be wrapped at 79 characters, even though PEP 8 suggests 72."_. :wink:  but I wrapped it differently to increase readability.
